src/mrpro/algorithms/reconstruction/DirectReconstruction.py

```diff
@@ -18,55 +17,56 @@ class DirectReconstruction(Reconstruction):
         """Initialize DirectReconstruction.
 
         Parameters
         ----------
+        kdata
+            KData. If kdata is provided and fourier_op or dcf are None, then fourier_op and dcf are estimated based on
+            kdata. Otherwise fourier_op and dcf are used as provided.
         fourier_op
-            Instance of the FourierOperator which adjoint is used for reconstruction.
+            Instance of the FourierOperator used for reconstruction. If None, set up based on kdata.
         csm
-            Sensitivity maps for coil combination. If None, no coil combination will be performed.
+            Sensitivity maps for coil combination. If a callable is provided, then kdata needs also to be provided.
         noise
-            Used for prewhitening
+            KNoise used for prewhitening. If None, no prewhitening is performed
         dcf
-            Density compensation. If None, no dcf will be performed.
-            Also set to None, if the FourierOperator is already density compensated.
+            K-space sampling density compensation. If None, set up based on kdata.
+
+        Raises
+        ------
+        ValueError
+            If the kdata and fourier_op are None or if csm is a Callable but kdata is None.
         """
         super().__init__()
-        self.fourier_op = fourier_op
-        # TODO: Make this buffers once DataBufferMixin is merged
-        self.dcf = dcf
-        self.csm = csm
-        self.noise = noise
+        if fourier_op is None:
+            if kdata is None:
+                raise ValueError('Either kdata or fourier_op needs to be defined.')
+            else:
+                self.fourier_op = FourierOp.from_kdata(kdata)
+        else:
+            self.fourier_op = fourier_op
 
-    @classmethod
-    def from_kdata(cls, kdata: KData, noise: KNoise | None = None, *, coil_combine: bool = True) -> Self:
-        """Create a DirectReconstruction from kdata with default settings.
+        if kdata is not None and dcf is None:
+            self.dcf = DcfData.from_traj_voronoi(kdata.traj)
+        else:
+            self.dcf = dcf
 
-        Parameters
-        ----------
-        kdata
-            KData to use for trajektory and header information.
-        noise
-            KNoise used for prewhitening. If None, no prewhitening is performed.
-        coil_combine
-            if True (default), uses kdata to estimate sensitivity maps and perform adaptive coil combine reconstruction
-            in the reconstruction.
-        """
-        if noise is not None:
-            kdata = prewhiten_kspace(kdata, noise)
-        dcf = DcfData.from_traj_voronoi(kdata.traj)
-        fourier_op = FourierOp.from_kdata(kdata)
-        self = cls(fourier_op, None, noise, dcf)
-        if coil_combine:
-            # kdata is prewhitened
-            self.recalculate_csm_walsh(kdata, noise=False)
-        return self
+        self.noise = noise
+
+        if csm is None or type(csm) is CsmData:
```

Comment:
this should be` isinstance(csm, CsmData)`
https://switowski.com/blog/type-vs-isinstance/

Reply:
this might already fix the mypy error